rfcs/0012-sequence-tracker.md

```diff
@@ -0,0 +1,174 @@
+#[allow(dead_code)]
+#[derive(PartialEq)]
+pub(crate) enum FindOption {
+    /// Round up to the next higher value when no exact match is found.
+    RoundUp,
+    /// Round down to the next lower value when no exact match is found.
+    RoundDown,
+}
+
+pub(crate) enum CliCommands {
+    // ...
+
+    /// Fetch an approximate (system) timestamp for a given sequence number.
+    /// SlateDB tracks a mapping between sequence numbers and timestamps and
+    /// maintains it with a lossy mechanism (over time the granularity of the
+    /// tracked mapping degrades).
+    GetTimestampForSeq {
+        /// The sequence number to fetch the timestamp for.
+        seq: u64,
+
+        /// The find option to use when fetching the timestamp.
+        find_opt: FindOption,
+    }
+
+    /// Fetch an approximate sequence number for a given timestamp.
+    /// SlateDB tracks a mapping between sequence numbers and timestamps and
+    /// maintains it with a lossy mechanism (over time the granularity of the
+    /// tracked mapping degrades).
+    GetSeqForTimestamp {
+        /// The timestamp to fetch the sequence number for.
+        ts: DateTime<Utc>,
+
+        /// The find option to use when fetching the sequence number.
+        find_opt: FindOption,
+    }
+}
+```
+
+### Manifest
+
+```fbs
+table ManifestV1 {
+    // ...
+
+    // The sequence tracker is a custom serialized type that is used to track
+    // a (lossy) mapping between sequence numbers and timestamps. The serialization
+    // format is described in RFC-0012.
+    seq_tracker: [ubyte];
+}
+```
+
+We choose to represent the sequence tracker as a custom serialized type instead
+of a flatbuffer type to reduce manifest storage overhead. It uses Gorilla
+encoding (delta-of-deltas) for both keys (sequence numbers) and values
+(timestamps).
+
+## Implementation
+
+The high-level approach for implementing the tracker is to store the mapping in
+the manifest and update it whenever a memtable is flushed or SlateDB cleanly
+shuts down.
+
+To reduce stored data we use two tactics:
+
+1. The encoding uses [Gorilla Encoding](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf),
+    which is an efficient encoding for timeseries data.
+2. The granularity is decimated over time (older seq-ts pairs are downsampled)
+
+### Serialization
+
+Physically, the sequence tracker is serialized as two Gorilla-encoded arrays of
+equal length (one for sequence numbers and one for timestamps). The arrays are
+broken up logically (though not physically) into tiers, and within a tier the
+timestamps expected to be evenly spaced (though this assumption is not
+enforced).
+
+The Gorilla encoding scheme means sequence numbers and timestamps will take up
+between 1 and 32 bits depending on SlateDB insertion regularity (typically
+closer to 7 bits each).
+
+The ultimate serialization format is:
+```
++-------------------------------------+
+| Version (u8) | Length (u32)         |
```

Comment:
What is version here?

Reply:
the version will let us evolve this if we ever want to do something fancier (like adding tiers), I was thinking it's worth the byte to avoid future backwards incompatible changes